File: 1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int index;
    private TreeNode buildBST(int upperBound, int[] preorder) {
        if(index == preorder.length || preorder[index] > upperBound) return null;
        TreeNode root = new TreeNode(preorder[index++]);
        root.left = buildBST(root.val, preorder);
        root.right = buildBST(upperBound, preorder);
        return root;
    }
    
    public TreeNode bstFromPreorder(int[] preorder) {
        index = 0;
        return  buildBST(Integer.MAX_VALUE, preorder);
    }
}

File: 1050-construct-binary-search-tree-from-preorder-traversal/construct-binary-search-tree-from-preorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int index;
    TreeNode* buildBST(vector<int>& preorder, int upperBound) 
    {
        if(index == preorder.size() || preorder[index] > upperBound)
        return nullptr;

        TreeNode* node = new TreeNode(preorder[index++]);
        
        node->left = buildBST(preorder, node->val);
        node->right = buildBST(preorder, upperBound);
        
        return node;
    }
    
public:
    TreeNode* bstFromPreorder(vector<int>& preorder) 
    {
        index = 0;
        return buildBST(preorder, INT_MAX);
    }
};